operator/cilium_node.go

```diff
@@ -1,112 +1,486 @@
+				key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+				if err != nil {
+					log.WithError(err).Warning("Unable to process CiliumNode Add event")
+					return
 				}
+				cfg.queue.Add(key)
 			},
 			UpdateFunc: func(oldObj, newObj interface{}) {
 				if oldNode := k8s.ObjToCiliumNode(oldObj); oldNode != nil {
-					if node := k8s.ObjToCiliumNode(newObj); node != nil {
-						if oldNode.DeepEqual(node) {
+					if newNode := k8s.ObjToCiliumNode(newObj); newNode != nil {
+						if oldNode.DeepEqual(newNode) {
+							return
+						}
+						key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(newObj)
+						if err != nil {
+							log.WithError(err).Warning("Unable to process CiliumNode Update event")
 							return
 						}
-						// node is deep copied before it is stored in pkg/aws/eni
-						nodeManager.Update(node)
+						cfg.queue.Add(key)
+					} else {
+						log.Warningf("Unknown CiliumNode object type %T received: %+v", newNode, newNode)
 					}
+				} else {
+					log.Warningf("Unknown CiliumNode object type %T received: %+v", oldNode, oldNode)
 				}
 			},
 			DeleteFunc: func(obj interface{}) {
-				if node := k8s.ObjToCiliumNode(obj); node != nil {
-					nodeManager.Delete(node.Name)
+				key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+				if err != nil {
+					log.WithError(err).Warning("Unable to process CiliumNode Delete event")
+					return
 				}
+				cfg.queue.Add(key)
 			},
-		},
-		k8s.ConvertToCiliumNode,
+		}
+	} else {
+		// Since we won't be handling any events we don't need to convert
+		// objects.
+		cfg.ciliumNodeConvertFunc = nil
+	}
+
+	return cfg, nil
+}
+
+// ciliumNodeName is only used to implement NamedKey interface.
+type ciliumNodeName struct {
+	name string
+}
+
+func (c *ciliumNodeName) GetKeyName() string {
+	return c.name
+}
+
+func startSynchronizingCiliumNodes(cfg *ciliumNodeWatcherCfg) error {
+	if cfg.withKVStore {
+		log.Info("Starting to synchronize CiliumNode custom resources to KVStore")
+	} else {
+		log.Info("Starting to synchronize CiliumNode custom resources")
+	}
+
+	// TODO: The operator is currently storing a full copy of the
+	// CiliumNode resource, as the resource grows, we may want to consider
+	// introducing a slim version of it.
+	var ciliumNodeInformer cache.Controller
+	ciliumNodeStore, ciliumNodeInformer = informer.NewInformer(
+		cache.NewListWatchFromClient(ciliumK8sClient.CiliumV2().RESTClient(),
+			cilium_v2.CNPluralName, v1.NamespaceAll, fields.Everything()),
+		&cilium_v2.CiliumNode{},
+		0,
+		cfg.resourceEventHandler,
+		cfg.ciliumNodeConvertFunc,
 	)
 
 	go func() {
 		cache.WaitForCacheSync(wait.NeverStop, ciliumNodeInformer.HasSynced)
 		close(k8sCiliumNodesCacheSynced)
+		// Only handle events if syncHandler is not nil. If it is nil then
+		// there isn't any event handler set for CiliumNodes events.
+		if cfg.syncHandler != nil {
+			for processNextWorkItem(cfg.queue, cfg.syncHandler) {
+			}
```

Comment:
I feel like this code is trying to trick me. At face value, the actual execution for each item that's processed is "do nothing", but actually the "do something" is a byproduct of the iterator. This code would be easier to follow if it were just opencoding the implementation like `for key, quit := cfg.queue.Get(); !quit { doStuff() }`.

Reply:
(On second read, I see that the actual execution is running `processNextWorkItem()` so it's clear that it will do something, but it's the empty braces that mislead me here.)